doc: add technical values document (#35145)
nodejs/node: doc/guides/technical-values.md

# Technical values and priorities

The project uses these technical values to establish priorities and guide
collaboration.

## Values and priority level

* Priority 1 - Developer experience
* Priority 2 - Stability
* Priority 3 - Operational qualities
* Priority 4 - Node.js maintainer experience

> * Priority 5 - Technology and API currency

Author: ```suggestion
* Priority 5 - Up to date Technology and APIs
```

### 2 - Stability
Whenever possible, we seek to insure that currently-working code continues to
work. We seek to keep the trust of developers and end-users. Therefore, we value
stability.
Some key elements of this include:
* Backwards compatibility
* Stable releases on a predictable schedule
* A strong safety net, including testing how changes
  in Node.js affect popular packages
* Careful consideration of what goes into long term support (LTS) releases

### 3 - Operational qualities
We value keeping Node.js safe, performant, and lightweight.
We value enabling the ability to investigate and debug problems in
development and production. Some key elements of this include:
* High throughput (speed)
* Fast startup
* Small binary size
* Small memory footprint
* High-quality debugging tools
* Robust diagnostic tools (profilers, etc.)
* Responsible security practices

### 4 - Node.js maintainer experience
We value the productivity and happiness of the Node.js maintainers.
Some key elements of this include:
* Approachability of the codebase
* Good internal documentation and guides
* Low-friction policies and processes
* Good CI and tooling to make maintainers productive

> ### 5 - Technology and API currency

Author: ```suggestion
### 5 - Up to date Technology and APIs
```

# Technical values and priorities

The project uses these technical values to establish priorities and guide
collaboration.

## Values and priority level

> This is only a subset of the overall priorities, listing only
> those which are at the top of the priority list. Anything on this
> list is very important for the project despite the relative priorities shown.
> 

Reviewer: I propose deleting this for the reasons previously stated: It dilutes rather than strengthens the values statement. And preventing questions along the lines of "what about X as a value?" is not (in my opinion) desirable.

```suggestion
```

* Priority 1 - Developer experience
* Priority 2 - Stability
* Priority 3 - Operational qualities
* Priority 4 - Node.js maintainer experience
* Priority 5 - Technology and API currency

## Value descriptions

### 1 - Developer experience
We value ensuring that developers are productive and enjoy developing
with Node.js. Some key elements of this include:
* Approachability (both technical and community)
* Great documentation
* Bundling friction-reducing APIs and components, even though
  they could be provided externally
* Enabling/supporting external packages to ensure overall developer experience

### 2 - Stability
Whenever possible, we seek to insure that currently-working code continues to

> work. We want to earn the trust of developers and end-users. Therefore, we value

Reviewer: Non-blocking question/suggestion (and I know I suggested the phrase in the first place):

```suggestion
work. We seek to keep the trust of developers and end-users. Therefore, we value
```

stability.
Some key elements of this include:
* Backwards compatibility
* Predictable and stable releases
* A strong safety net, including testing how changes
  in Node.js affect popular packages
* Careful consideration of what goes into long term support (LTS) releases

### 3 - Operational qualities
We value keeping Node.js safe, performant, and lightweight.
We value enabling the ability to investigate and debug problems in
development and production. Some key elements of this include:
* High throughput (speed)
* Fast startup
* Small binary size
* Small memory footprint
* High-quality debugging tools

> * Diagnostic tooling (profilers, heapdumps, coredumps, etc.)

Reviewer: I don't think of heapdumps and coredumps as tools. They're data consumed by tools like profilers and debuggers. Maybe this?:

```suggestion
* Robust diagnostic tools (profilers, etc.)
```
